src/GitRepository.cc

```diff
@@ -0,0 +1,415 @@
+            else
+                filestats.path_name = old_path ? std::string{old_path} : std::string{new_path};
+
+            return_array.push_back(filestats);
+
+            continue;
+        }
+
+        // list files which are staged for next commit
+        //############################################
+
+        if (s->status & GIT_STATUS_INDEX_NEW)
+            istatus = "new file";
+        if (s->status & GIT_STATUS_INDEX_MODIFIED)
+            istatus = "modified";
+        if (s->status & GIT_STATUS_INDEX_DELETED)
+            istatus = "deleted";
+        if (s->status & GIT_STATUS_INDEX_RENAMED)
+            istatus = "renamed";
+        if (s->status & GIT_STATUS_INDEX_TYPECHANGE)
+            istatus = "typechange";
+
+        if (istatus != "")
+        {
+          
+            filestats.handling = "staged";
+            filestats.changes = std::string{istatus};
+
+            old_path = s->head_to_index->old_file.path;
+            new_path = s->head_to_index->new_file.path;
+
+            if (old_path && new_path && strcmp(old_path, new_path))
+                filestats.path_name = gul14::cat(old_path, " -> ", new_path);
+            else
+                filestats.path_name = old_path ? std::string{old_path} : std::string{new_path};
+
+            return_array.push_back(filestats);
+
+            
+            continue;
+        }
+
+        // list untracked files
+        //######################
+        if (s->status == GIT_STATUS_WT_NEW)
+        {
+
+            filestats.handling = "untracked";
+            filestats.changes = "untracked";
+            filestats.path_name = std::string(s->index_to_workdir->old_file.path);
+
+            return_array.push_back(filestats);
+
+            continue;
+        }
+
+        // list ignored files
+        //####################
+        if (s->status == GIT_STATUS_IGNORED) {
+
+            filestats.handling = "ignored";
+            filestats.changes = "ignored";
+            filestats.path_name = std::string(s->index_to_workdir->old_file.path);
+
+            return_array.push_back(filestats);
+
+            continue;
+        }
+
+        delete [] old_path;
+        delete [] new_path;
+    }
+    return return_array;
+}
+
+std::vector<FileStatus> GitRepository::status()
+{
+    // declare necessary variables
+    git_status_list *my_status;
+    git_status_options status_opt;
+
+    // init options with default values (= no args)
+    int error = git_status_init_options(&status_opt, GIT_STATUS_OPT_INCLUDE_UNTRACKED);// & GIT_STATUS_OPT_INCLUDE_IGNORED);
+    if (error) throw task::Error("Cannot init status options.");
+
+    // fill C-type status pointer
+    error = git_status_list_new(&my_status, repo_.get(), &status_opt);
+    if (error) throw task::Error("Cannot init status.");
+
+    // translate status pointer to redable status information
+    std::vector<FileStatus> status_arr = collect_status(my_status);
+
+    git_status_list_free(my_status);
```

Comment:
Unfortunately `collect_status()` might throw – it is somewhat unlikely, but still. If that happens, `my_status` will never be freed. So it should be wrapped in a `LibGitPointer`.

Reply:
Another libgit pointer I oversaw...